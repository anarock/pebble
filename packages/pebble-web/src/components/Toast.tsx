import * as React from "react";
import { toastWrapper } from "./styles/Toast.styles";
import { colors } from "pebble-shared";
import {
  ToastProps,
  ToastState,
  ToastType,
  ToastPosition
} from "./typings/Toast";
import { Transition, animated } from "react-spring/renderprops.cjs";

import mitt from "mitt";
import { animationConfig } from "../utils/animation";

const emitter = /*#__PURE__*/ mitt();

export const _colors = {
  success: colors.emerald.base,
  error: colors.red.base
};

const toastTransitionsLeft = {
  from: { transform: "translateX(-10px)" },
  enter: { transform: "translateX(0)" },
  leave: { transform: "translateX(-10px)" }
};

const toastTransitionsRight = {
  from: { transform: "translateX(10px)" },
  enter: { transform: "translateX(0)" },
  leave: { transform: "translateX(10px)" }
};

const customStyles = {
  TOP: {
    style: {
      top: 20,
      left: "50%"
    },

    transitions: {
      from: { transform: "translateX(-50%) translateY(-10px)" },
      enter: { transform: "translateX(-50%) translateY(0)" },
      leave: { transform: "translateX(-50%) translateY(-10px)" }
    }
  },

  TOP_LEFT: {
    style: {
      top: 20,
      left: 20
    },

    transitions: toastTransitionsLeft
  },

  TOP_RIGHT: {
    style: {
      top: 20,
      right: 20
    },

    transitions: toastTransitionsRight
  },

  BOTTOM: {
    style: {
      bottom: 20,
      left: "50%"
    },

    transitions: {
      from: { transform: "translateX(-50%) translateY(10px)" },
      enter: { transform: "translateX(-50%) translateY(0)" },
      leave: { transform: "translateX(-50%) translateY(10px)" }
    }
  },

  BOTTOM_LEFT: {
    style: {
      bottom: 20,
      left: 20
    },

    transitions: toastTransitionsLeft
  },

  BOTTOM_RIGHT: {
    style: {
      bottom: 20,
      right: 20
    },

    transitions: toastTransitionsRight
  }
};

class Toast extends React.PureComponent<ToastProps, ToastState> {
  static show(
    text: string,
    type: ToastType,
    { time, position }: { time?: number; position?: ToastPosition } = {}
  ) {
    emitter.emit("showToast", { text, type, time, position });
  }

  showTimer?: number | null;

  static hide() {
    emitter.emit("hideToast");
  }

  state: ToastState = {
    text: "",
    type: "success",
    show: false,
    position: "BOTTOM"
  };

  componentDidMount() {
    emitter.on("showToast", this.show);
    emitter.on("hideToast", this.hide);
  }

  componentWillUnmount() {
    emitter.off("showToast", this.show);
    emitter.off("hideToast", this.hide);
  }

  private show = ({
    text,
    type = "success",
    position,
    time
  }: Partial<ToastState> & {
    text: string;
    time?: number;
    position: ToastPosition;
  }) => {
    this.setState({
      text,
      type,
      position,
      show: true
    });

    if (this.showTimer) {
      clearTimeout(this.showTimer);
      this.showTimer = null;
    }

    this.showTimer = window.setTimeout(
      () =>
        this.setState({
          show: false
        }),
      time ? time : this.props.defaultTime || 5000
    );
  };

  private hide = () => this.setState({ show: false });

  render() {
    const bColor = _colors[this.state.type];

    const iconClass = `pi ${(this.state.type === "success" &&
      "pi-radio-check-filled") ||
      ""} ${this.state.type === "error" && "pi-close-circle-filled"}`;

    const position =
      this.state.position || this.props.defaultPosition || "BOTTOM";

    return (
      <Transition
        native
        items={this.state.show}
        key={position}
        from={{
          opacity: 0,
          ...customStyles[position].transitions.from
        }}
        enter={{
          opacity: 1,
          ...customStyles[position].transitions.enter
        }}
        leave={{
          opacity: 0,
          pointerEvents: "none",
          ...customStyles[position].transitions.leave
        }}
        config={animationConfig.config}
      >
        {show =>
          show &&
          (styles => (
            <animated.div
              css={[toastWrapper, this.props.styles]}
              style={{
                backgroundColor: bColor,
                ...(styles as React.CSSProperties),
                ...customStyles[position].style
              }}
            >
              <i className={iconClass} />
              {this.state.text}
            </animated.div>
          ))
        }
      </Transition>
    );
  }
}

export default Toast;
