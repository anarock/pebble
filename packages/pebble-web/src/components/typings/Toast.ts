import { Interpolation } from "./emotionCustom";

export type ToastType = "success" | "error";

export type ToastPosition =
  | "BOTTOM"
  | "BOTTOM_LEFT"
  | "BOTTOM_RIGHT"
  | "TOP"
  | "TOP_LEFT"
  | "TOP_RIGHT";

export interface ToastProps {
  styles?: Interpolation;
  defaultTime?: number;
  defaultPosition?: ToastPosition;
}

export interface ToastState {
  type: ToastType;
  text: string;
  show: boolean;
  position?: ToastPosition;
}
