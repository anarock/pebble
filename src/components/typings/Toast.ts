export type ToastType = "success" | "error";

export enum ToastPosition {
  BOTTOM = "BOTTOM",
  RIGHT = "RIGHT"
}

export interface ToastProps {
  className?: string;
  defaultTime?: number;
}

export interface ToastState {
  type: ToastType;
  text: string;
  show: boolean;
  position: ToastPosition;
}
