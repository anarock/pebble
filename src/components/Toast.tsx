import * as React from "react";
import { toastWrapper } from "./styles/Toast.styles";
import { colors } from "../theme";
import { ToastState, ToastType } from "./typings/Toast";
import { Transition, animated } from "react-spring";
import { cx } from "emotion";
import Mitt from "mitt";

const emitter = new Mitt();

const _colors = {
  success: colors.emerald.base,
  error: colors.red.base
};

class Toast extends React.PureComponent<{}, ToastState> {
  static show(text: string, type: ToastType) {
    emitter.emit("showToast", { text, type });
  }

  static hide() {
    emitter.emit("hideToast");
  }

  state = {
    text: "",
    type: "success",
    show: false
  };

  componentDidMount() {
    emitter.on("showToast", this.show);
    emitter.on("hideToast", this.hide);
  }

  componentWillUnmount() {
    emitter.off("showToast", this.show);
    emitter.off("hideToast", this.hide);
  }

  private show = ({ text, type = "success" }: Partial<ToastState>) => {
    this.setState({
      text,
      type,
      show: true
    });

    setTimeout(
      () =>
        this.setState({
          show: false
        }),
      2000
    );
  };

  private hide = () => this.setState({ show: false });

  render() {
    const bColor = _colors[this.state.type];

    const iconClass = cx({
      "icon-radio-check-filled": this.state.type === "success",
      "icon-close-circle-filled": this.state.type === "error"
    });

    return (
      // @ts-ignore
      <Transition
        from={{ opacity: 0, transform: "translateX(-50%) translateY(10px)" }}
        enter={{ opacity: 1, transform: "translateX(-50%) translateY(0)" }}
        leave={{ opacity: 0, transform: "translateX(-50%) translateY(10px)" }}
      >
        {this.state.show &&
          (styles => (
            <animated.div
              className={toastWrapper}
              style={{ backgroundColor: bColor, ...styles }}
            >
              <i className={iconClass} />
              {this.state.text}
            </animated.div>
          ))}
      </Transition>
    );
  }
}

export default Toast;
