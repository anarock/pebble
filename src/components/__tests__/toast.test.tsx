import * as React from "react";
import { mount } from "enzyme";
import Toast from "../Toast";
import sinon from "sinon";

describe("Component: Toast", () => {
  test("snapshot", () => {
    const toast = mount(<Toast />);

    toast.setState({
      show: true,
      text: "hello world",
      type: "success"
    });

    expect(toast).toMatchSnapshot();
  });

  test("renders when called and disappears automatically.", () => {
    const clock = sinon.useFakeTimers();

    const toast = mount(<Toast />);

    Toast.show("Hello there", "success");

    expect(toast.state()).toEqual({
      text: "Hello there",
      show: true,
      type: "success"
    });

    clock.tick(2000);
    expect(toast.state("show")).toBeFalsy();
  });

  test("manually hide toast", () => {
    const toast = mount(<Toast />);

    Toast.show("Hello there", "success");
    expect(toast.state("show")).toBeTruthy();

    Toast.hide();
    expect(toast.state("show")).toBeFalsy();
  });
});
