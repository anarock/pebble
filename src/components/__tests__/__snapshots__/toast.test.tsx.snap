// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Toast snapshot 1`] = `
Array [
  "",
  .emotion-0 {
  position: fixed;
  bottom: 50px;
  border-radius: 3px;
  font-weight: 400;
  color: #ffffff;
  font-size: 14px;
  height: 46px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 20px;
  z-index: 99999;
  left: 50%;
}

.emotion-0 > i {
  margin-right: 10px;
}

<ForwardRef
    className="emotion-0"
    style={
      Object {
        "backgroundColor": "#10AC60",
        "opacity": 0,
        "originalKey": "styles => {
                    /* istanbul ignore next */cov_3rzavxems.f[9]++;
                    cov_3rzavxems.s[27]++;
                    return (/* istanbul ignore next */React.createElement(
                            /* istanbul ignore next */react_spring_1.animated.div,
                            /* istanbul ignore next */{ className: Toast_styles_1.toastWrapper, style: /* istanbul ignore next */_extends({ backgroundColor: bColor }, styles) },
                            /* istanbul ignore next */React.createElement( /* istanbul ignore next */\\"i\\", /* istanbul ignore next */{ className: iconClass }),
                            this.state.text
                        )
                    );
                }",
        "transform": "translateX(-50%) translateY(10px)",
      }
    }
  >
    <div
      className="emotion-0"
      style={
        Object {
          "backgroundColor": "#10AC60",
          "opacity": 0,
          "originalKey": "styles => {
                    /* istanbul ignore next */cov_3rzavxems.f[9]++;
                    cov_3rzavxems.s[27]++;
                    return (/* istanbul ignore next */React.createElement(
                            /* istanbul ignore next */react_spring_1.animated.div,
                            /* istanbul ignore next */{ className: Toast_styles_1.toastWrapper, style: /* istanbul ignore next */_extends({ backgroundColor: bColor }, styles) },
                            /* istanbul ignore next */React.createElement( /* istanbul ignore next */\\"i\\", /* istanbul ignore next */{ className: iconClass }),
                            this.state.text
                        )
                    );
                }",
          "transform": "translateX(-50%) translateY(10px)",
        }
      }
    >
      <i
        className="icon-radio-check-filled"
      />
      hello world
    </div>
  </ForwardRef>,
]
`;
